Build a regression model.

In [8]:
# imports
import pandas as pd
import statsmodels.api as sm

grouped_business_df = pd.read_csv('grouped_business.csv', index_col = 0)

In [9]:
grouped_business_df

Station Name  Average Rating  Average Reviews  \
0            Broadway Residence Hall        4.175000       572.050000   
1                Casey Eye Institute        4.175000        80.850000   
2              Cleveland High School        4.300000       561.050000   
3                         Cully Park        4.083333        47.111111   
4    Doernbecher Children's Hospital        4.175000        48.750000   
..                               ...             ...              ...   
233      SW Yamhill at Director Park        4.175000      1497.900000   
234                    Shattuck Hall        4.175000       605.500000   
235         Tilikum West at SW Moody        4.025000       285.600000   
236     Urban Center Plaza at SW 6th        4.150000       741.550000   
237        W Burnside at SW Park Ave        4.200000      1707.550000   

     Number of Bikes  
0                  4  
1                  6  
2                  9  
3                 10  
4                  6  
..               ...  
233               20  
234                3  
235               18  
236               14  
237                9  

[238 rows x 4 columns]

In [10]:
X = grouped_business_df['Number of Bikes']
Y = grouped_business_df['Average Rating']

In [11]:
X = sm.add_constant(X)
lin_model = sm.OLS(Y, X).fit()

In [12]:
model = lin_model.summary()

Provide model output and an interpretation of the results. 

In [13]:
print(model)

                            OLS Regression Results                            
Dep. Variable:         Average Rating   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.004
Method:                 Least Squares   F-statistic:                   0.02730
Date:                Tue, 12 Dec 2023   Prob (F-statistic):              0.869
Time:                        22:17:59   Log-Likelihood:                 88.758
No. Observations:                 238   AIC:                            -173.5
Df Residuals:                     236   BIC:                            -166.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const               4.2146      0.023    1

In [ ]:
#Inference from above statsmodel:
#R-squared measures the proportion of the variance in the dependent variable (Average Rating) explained by the independent variable(s) (Number of Bikes).
#In this case, R-squared is 0.000, indicating that the model does not explain a significant amount of variability in 'Average Rating'. The model doesn't fit the data well and is not statistically significant
#Also the p-value for the Number of Bikes is actually very high at 0.869, we typically are looking for less than 0.05.

In [15]:
#using 'Average Reviews' for the stats model
X = grouped_business_df['Number of Bikes']
Y = grouped_business_df['Average Reviews']

X = sm.add_constant(X)
lin_model = sm.OLS(Y, X).fit()
model = lin_model.summary()
print(model)

                            OLS Regression Results                            
Dep. Variable:        Average Reviews   R-squared:                       0.137
Model:                            OLS   Adj. R-squared:                  0.133
Method:                 Least Squares   F-statistic:                     37.43
Date:                Tue, 12 Dec 2023   Prob (F-statistic):           3.90e-09
Time:                        22:33:09   Log-Likelihood:                -1760.3
No. Observations:                 238   AIC:                             3525.
Df Residuals:                     236   BIC:                             3532.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const             274.1422     55.022     

Interpretation of Results:

R-squared: From the model above, an R-squared value of 0.137 indicates that about 13.7% of the variability in the dependent variable (Average Reviews) is explained by the variability in the independent variable (Number of Bikes). Thus, this model suggests a moderate level of variation in the dependent variable.

P-value: In this model output, the P-value for the Number of Bikes is 0.000. The low p-value suggests that we can reject the null hypothesis. The p-value indicates that the 'Number of Bikes' variable is a significant predictor of 'Average Reviews' in the model. 'Number of Bikes' variable has statistically significant effect on 'Average Reviews'.

Coefficient: For this linear regression model, the coefficient for Number of Bikes is approximately 30.8923, suggesting an estimated change in 'Average Reviews' for a one-unit increase in the 'Number of Bikes'.

F-statistic: In this model output, the F-statistic is 37.43 with a very low p-value (3.90e-09) which is an indicator that the model is statistically significant, and the 'Number of Bikes' is a significant predictor of 'Average Reviews'.

Conclusion: While we can establish a statistically significant relationship between the independent and dependent variables, it is also worthy to note that the relationship between them does not imply causation.


# Stretch

How can you turn the regression model into a classification model?

To convert the regression model into a classification model, I will first define which attribute I am trying to predict. Next, I'll work on grouping my continuous variables into discrete labels to help create the class. Then I'll select an approrpiate classification algorithm and afterwards, evaluate my model.